### PR TITLE
[TIR, TVMScript] Update printer / parser to make T.allocate return buffer var

### PR DESCRIPTION
**Pull request by vinx13**
* Updated TVMScript syntax of `T.allocate` to return buffer var.
* Added syntax sugar for `T.decl_buffer`. When `data` field is not specified, `data` will be implicitly created via `Allocate` stmt.
* Updated the existing test cases. Most test cases can be updated by changing `T.allocate` to `T.decl_buffer`. `T.allocate` in some tests are updated to `T.allocate` + `T.buffer_decl`, to maintain the legacy behavior of allocation and implicit buffer declaration (will be followed up in future PR to adopt `T.decl_buffer`).

Related RFC: https://github.com/apache/tvm-rfcs/pull/87

cc @junrushao1994 @Lunderberg @wrongtest-intellif @cyx-6 @tqchen

### CODE REVIEW

**Review by Lunderberg — COMMENTED** [review]
Looks good!  I have a couple of questions on the implementation side, but overall it looks very good!

File: tests/python/unittest/test_tir_transform_unroll_loop.py [hunk 1]
```diff
@@ -117,16 +117,19 @@ class before:
...
+                with T.allocate([16], "float32", "global") as buf_data:
+                    buf = T.buffer_decl(shape=[16], dtype="float32", scope="global", data=buf_data)
                     buf[0] = 0.0
 
     @tvm.script.ir_module
     class expected:
         @T.prim_func
         def main():
-            with T.allocate([16], "float32", "global") as buf1:
+            with T.allocate([16], "float32", "global") as buf1_data:
```
[the remark this answers is not in the document]
Comment (vinx13) [reply tests/python/unittest/test_tir_transform_unroll_loop.py:128]: Yes. The reason I kept `T.allocate` here is that the pass also need some updates before we can use `T.decl_buffer`. (The scope of this PR is to only update existing TVM scripts without touching related passes to minimize changes)

File: src/printer/tvmscript_printer.cc [hunk 2]
```diff
@@ -1028,59 +1039,52 @@ Doc TVMScriptPrinter::VisitStmt_(const BufferRealizeNode* op) {
...
-  Buffer alloc_buffer;
-  Array<Buffer> aliasing_buffers;
-};
-
-template <typename AllocNode>
-AllocUsage FindAllocateUsage(AllocNode* op, Map<Var, Array<Buffer>>* cache_ptr) {
-  Map<Var, Array<Buffer>>& cache = *cache_ptr;
-  if (!cache.count(op->buffer_var)) {
-    cache = BufferUsageFinder::FindUsage(std::move(cache), op->body);
+bool IsAllocateDeclBufferPattern(const AllocateNode* allocate, const DeclBufferNode* decl_buffer) {
```
[the remark this answers is not in the document]
Comment (vinx13) [reply src/printer/tvmscript_printer.cc:1042]: Nice catch! Indeed it's clearer without `decl_buffer` argument

File: src/printer/tvmscript_printer.cc [hunk 3]
```diff
@@ -1028,59 +1039,52 @@ Doc TVMScriptPrinter::VisitStmt_(const BufferRealizeNode* op) {
...
-  buf_not_in_headers_.insert(alloc_buffer.get());
-  var_not_in_headers_.insert(alloc_buffer->data.get());
+  var_not_in_headers_.insert(op->buffer_var.get());
+
+  if (!buffer_var_usage_.count(op->buffer_var)) {
+    buffer_var_usage_ = BufferUsageFinder::FindUsage(std::move(buffer_var_usage_), op->body);
+  }
+  Array<Buffer> buffer_usage = buffer_var_usage_.Get(op->buffer_var).value_or({});
+
+  if (buffer_usage.empty()) {
```
[the remark this answers is not in the document]
Comment (vinx13) [reply src/printer/tvmscript_printer.cc:1077]: Usage in `DeclBuffer` is excluded from the result of `BufferUsageFinder`

File: src/printer/tvmscript_printer.cc [hunk 4]
```diff
@@ -1028,59 +1039,52 @@ Doc TVMScriptPrinter::VisitStmt_(const BufferRealizeNode* op) {
...
+    if (const DeclBufferNode* decl_buffer = op->body.as<DeclBufferNode>()) {
+      if (IsAllocateDeclBufferPattern(op, decl_buffer)) {
+        // As a syntax sugar, we identify the pattern of Allocate and DeclBuffer and print a single
+        // DeclBuffer statement. It is intentionally to call `Print` instead of `PrintBody` here to
+        // delegate the printing of the current node to `DeclBufferNode` while maintaining the
+        // same value of `current_num_` and `num_child_`.
+        return Print(op->body);
```
[the remark this answers is not in the document]
Comment (vinx13) [reply src/printer/tvmscript_printer.cc:1084]: That's correct. So I checked `buffer_usage.empty()` above to make sure `T.buffer_decl` is not needed

**Review by vinx13 — COMMENTED** [review]
To answer the question about `T.allocate` vs `T.decl_buffer` update:
There are two ways to update the existing TVM scripts.
1. Update `T.allocate` to `T.allocate` + `T.buffer_decl`. This is strictly equivalent to the TIR before.
2. Update `T.allocate` to `T.allocate + `T.decl_buffer`, (or the syntax sugar form, only a single `T.decl_buffer` statement). This adds a `DeclBuffer` node in TIR, which is not strictly equivalent to the TIR before. This is the preferred way (as it uses explicit `DeclBuffer`) but might require updating the existing passes. To minimize changes to existing passes in this PR, we will choose method 1 for now.

File: src/printer/tvmscript_printer.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (Lunderberg) [reply src/printer/tvmscript_printer.cc:1084]: Thank you, and that makes sense.  I had been picturing the allocate/decl syntax as being used whenever it appears, where it is instead restricted to only cases where there is an allocate/decl and there is no other aliasing of that buffer.  If there is aliasing, then we don't necessarily have the cleanest tvmscript that is printed out, but that doesn't really matter with the `decl_buffer` node is on its way to being mandatory.

Marking both this question and the one about `buffer_usage.empty()` as resolved.

**Comment by Lunderberg** [comment]
That makes sense for minimizing the changes to the generated TIR, so that those can be handled in an independent change.  Marking all of those conversations as resolved.

**Review by Lunderberg — APPROVED** [review]
LGTM!

**Pull request merged**